sertop/sername.ml

```diff
@@ -0,0 +1,211 @@
+      ; deep_edits    = false
+      ; async_workers = async_workers
+      } in
+
+  (* Disable due to https://github.com/ejgallego/coq-serapi/pull/94 *)
+  let stm_options =
+    { stm_options with
+      async_proofs_tac_error_resilience = `None
+    ; async_proofs_cmd_error_resilience = false
+    } in
+
+  let stm_options =
+    if quick
+    then { stm_options with async_proofs_mode = APonLazy }
+    else stm_options
+  in
+
+  (*
+  let require_libs =
+    let prelude = ["Coq.Init.Prelude", None, Some false] in
+    match require_lib with
+    | Some l -> prelude @ [l, None, Some false]
+    | None -> prelude
+  in
+  *)
+  let require_libs = ["Coq.Init.Prelude", None, Some false] in
+
+  let ndoc = { Stm.doc_type = Stm.VoDoc in_file
+             ; require_libs
+             ; iload_path
+             ; stm_options
+             } in
+
+  (* Workaround, see
+     https://github.com/ejgallego/coq-serapi/pull/101 *)
+  if quick || async <> None
+  then Safe_typing.allow_delayed_constants := true;
+
+  match require_lib with
+  | None -> Stm.new_doc ndoc
+  | Some l ->
+     (*
+     let sdoc = Stm.new_doc ndoc in
+     let dir,from,exp = l,None,Some false in
+     let mp = Libnames.qualid_of_string dir in
+     let mfrom = Option.map Libnames.qualid_of_string from in
+     Flags.silently (Vernacentries.vernac_require mfrom exp) [mp];
+     sdoc
+     *)
+     let doc,sid = Stm.new_doc ndoc in
+     let sent = Printf.sprintf "Require %s." l in
+     let in_strm = Stream.of_string sent in
+     let in_pa = Pcoq.Parsable.make ~loc:(Loc.initial (InFile in_file)) in_strm in
+     match Stm.parse_sentence ~doc ~entry:Pvernac.main_entry sid in_pa with
+     | Some ast ->
+	let doc, sid, tip = Stm.add ~doc ~ontop:sid false ast in
+	if tip <> `NewTip then CErrors.user_err ?loc:ast.loc Pp.(str "fatal, got no `NewTip`");
+	doc, sid
+     | None -> assert false
+
+exception End_of_input
+
+let input_doc ~in_chan ~process ~doc ~sid =
+  try while true; do
+      let line = input_line in_chan in
+      if String.trim line <> "" then process ~doc ~sid line
+    done
+  with End_of_file -> ()
+
+let context_of_st m = match m with
+  | `Valid (Some { Vernacstate.proof = Some pstate; _ } ) ->
+    Pfedit.get_current_context pstate
+  | _ ->
+    let env = Global.env () in Evd.from_env env, env
+
+let str_pp_obj env sigma fmt (obj : Serapi_protocol.coq_object) : unit =
+  Format.fprintf fmt "%a" Pp.pp_with (Serapi_protocol.gen_pp_obj env sigma obj)
+
+let process_line ~pp ~str_pp ~de_bruijn ~body ~doc ~sid line =
+  let open Serapi_protocol in
+  let st = Stm.state_of_id ~doc sid in
+  let sigma, env = context_of_st st in
+  let info = QueryUtil.info_of_id env line in
+  let def = if body then fst info else snd info in
+  match def with
+  | [CoqConstr def_term] ->
+     let evd = Evd.from_env env in
+     let edef_term = EConstr.of_constr def_term in
+     let gdef_term = Detyping.detype Detyping.Now false Names.Id.Set.empty env evd edef_term in
```

Comment:
> Thinking a bit about it, I propose we mark `sername` deprecated in 8.11/8.12 once can be done with Query, but don't remove it, at least until it becomes a problem for maintenance.

This is also perfectly fine with me. However, unless I'm missing something, I think the main obstacle to turning `sername` into a simple query separate from Coq's API is the  use of detyping here. For example, would this best be done by adding another query function?